cray_product_catalog/catalog_delete.py

```diff
@@ -101,10 +113,10 @@ def modify_config_map(name, namespace, product, product_version, key=None):
             LOGGER.exception("Error calling read_namespaced_config_map")
 
             # ConfigMap doesn't exist yet
-            if err.status == 404 and attempt < max_attempts:
-                LOGGER.warning("ConfigMap %s/%s doesn't exist, attempting again.", namespace, name)
-                continue
-            raise  # unrecoverable
+            if err.status != ERR_NOT_FOUND and attempt > max_attempts:
+                raise   # unrecoverable
+            LOGGER.warning("ConfigMap %s/%s doesn't exist, attempting again.", namespace, name)
+            continue
```

Comment:
It appears this new logic is not equivalent to the old logic.

This used to raise an exception if the error is anything other than a 404 or if the error is a 404, and it has exceeded the maximum number of attempts.

The new logic will raise an exception if the error is anything other than a 404 _and_ the number of attempts is exceeded (actually exceeded by 1 due to an off-by-one-error in negating the less-than operator). The case where this behavior differs is when the `err.status` is something other than 404 (e.g. a 409 conflict). Under the old logic, it would never try again for such an error. Under the new logic, it will try again up to `max_attempts + 1` even though it may be hopeless or incorrect to do so.

Put another way, let `A` represent `err.status == 404`, and let `B` represent `attempt < max_attempts`. Then the old logic was:

```
if A and B:
    LOGGER.warning('...')
    continue
else:
    raise
```

The new logic is (ignoring the fact that the less-than operator is incorrectly negated):

```
if not A and not B:
    raise
else:
    LOGGER.warning('...')
    continue
```

The correct way to negate `A and B` is either `not (A and B)` or `(not A) or (not B)` by [De Morgan's laws](https://en.wikipedia.org/wiki/De_Morgan%27s_laws). Expanding the latter alternative, we get:

```
if err.status != 404 or attempt >= max_attempts:
    raise
else:
    LOGGER.warning('...')
    continue
```

Having said all that, I'm wondering why attempt to change the order of the consequent and alternative in the if-then statement. Is there a reason I'm missing?

Reply:
Too many changes we did around this. Later we simplified and the logic was overlooked. Old code would work so reverting it back to the actual code